rclpy/src/rclpy/_rclpy.c

```diff
@@ -2221,141 +2294,32 @@ rclpy_service_server_is_available(PyObject * Py_UNUSED(self), PyObject * args)
   Py_RETURN_FALSE;
 }
 
-/// Destroy an entity attached to a node
-/**
- * Entity type must be one of ["publisher", "client", "service"].
- *
- * Raises RuntimeError on failure
- *
- * \param[in] pyentity Capsule pointing to the entity to destroy
- * \param[in] pynode Capsule pointing to the node the entity belongs to
- */
-static PyObject *
-rclpy_destroy_node_entity(PyObject * Py_UNUSED(self), PyObject * args)
-{
-  PyObject * pyentity;
-  PyObject * pynode;
-
-  if (!PyArg_ParseTuple(args, "OO", &pyentity, &pynode)) {
-    return NULL;
-  }
-
-  rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(pynode, "rcl_node_t");
-  if (!node) {
-    return NULL;
-  }
-
-  if (!PyCapsule_CheckExact(pyentity)) {
-    PyErr_Format(PyExc_RuntimeError, "entity is not a capsule");
-    return NULL;
-  }
-
-  rcl_ret_t ret;
-  if (PyCapsule_IsValid(pyentity, "rcl_publisher_t")) {
-    rcl_publisher_t * publisher = (rcl_publisher_t *)PyCapsule_GetPointer(
-      pyentity, "rcl_publisher_t");
-    ret = rcl_publisher_fini(publisher, node);
-    PyMem_Free(publisher);
-  } else if (PyCapsule_IsValid(pyentity, "rcl_client_t")) {
-    rcl_client_t * client = (rcl_client_t *)PyCapsule_GetPointer(pyentity, "rcl_client_t");
-    ret = rcl_client_fini(client, node);
-    PyMem_Free(client);
-  } else if (PyCapsule_IsValid(pyentity, "rcl_service_t")) {
-    rcl_service_t * service = (rcl_service_t *)PyCapsule_GetPointer(pyentity, "rcl_service_t");
-    ret = rcl_service_fini(service, node);
-    PyMem_Free(service);
-  } else {
-    ret = RCL_RET_ERROR;  // to avoid a linter warning
-    PyErr_Format(PyExc_RuntimeError, "'%s' is not a known node entity",
-      PyCapsule_GetName(pyentity));
-    return NULL;
-  }
-  if (ret != RCL_RET_OK) {
-    PyErr_Format(PyExc_RuntimeError,
-      "Failed to fini '%s': %s", PyCapsule_GetName(pyentity), rcl_get_error_string().str);
-    rcl_reset_error();
-    return NULL;
-  }
-
-  if (PyCapsule_SetPointer(pyentity, Py_None)) {
-    // exception set by PyCapsule_SetPointer
-    return NULL;
-  }
-
-  Py_RETURN_NONE;
-}
-
 /// Destructor for a clock
-void
+static void
 _rclpy_destroy_clock(PyObject * pycapsule)
 {
   rcl_clock_t * clock = (rcl_clock_t *)PyCapsule_GetPointer(pycapsule, "rcl_clock_t");
   if (NULL == clock) {
```

Comment:
I know you didn't add this, but for consistency it would probably be better to use `!clock` like all of the rest of the PyCapsulte_GetPointer checks.

Reply:
It looks like `_rclpy` is inconsistent about `NULL == foo` vs `!foo` after `PyCapsule_GetPointer`.
https://github.com/ros2/rclpy/blob/1b4728b8bc75d5e8ae765e938ed2513d24f7e45d/rclpy/src/rclpy/_rclpy.c#L655-L656
https://github.com/ros2/rclpy/blob/1b4728b8bc75d5e8ae765e938ed2513d24f7e45d/rclpy/src/rclpy/_rclpy.c#L4457-L4458

 After jumping through the calls it feels like `!foo` is more popular, but there are a lot of `NULL== foo`. If it's alright, I'd like to open an issue instead changing them in this PR.

Reply:
Opened #331